Allow more explicit control of the LFN and PFN.  Fixes #377

bbockelm opened the pull request:
This allows the upload path to track the "file name" (on local disk) and the LFN separately, allowing the `rucio upload` client to expose the option of explicitly providing a LFN that differs from the file
name.

Additionally, to help upload files with '/' in them that are effectively pre-hashed, this allows the RSE to be marked as having the 'identity' determinism type, meaning to skipping the hashing step of generating a PFN.

This is the `next` equivalent of #378.

vingar commented:
bin/rucio:934:21: E225 missing whitespace around operator
bin/rucio:1153:56: E127 continuation line over-indented for visual indent
lib/rucio/rse/rsemanager.py:343:4: W291 trailing whitespace
lib/rucio/rse/rsemanager.py:346:12: W291 trailing whitespace

bari12 commented:
There are some implications of the determinism_type which we have to check, because this reaches much further than just the protocol as it reaches into the transfers too. But I have some basic questions to this as well, will post them to the issue.

vingar commented:
there are some comments on the issue as well. We need to discuss and think a bit here

bari12 commented on bin/rucio:
```diff
@@ -918,12 +918,26 @@ def upload(args):
+        logger.error("A single GUID was specified on the command line, but there are multiple files to upload.")
+        logger.error("If GUID auto-detection is not used, only one file may be uploaded at a time")
+        return FAILURE
+    if len(files) > 1 and args.lfn:
```
To be consistent with rucio terminology `lfn` should be renamed to `name`. And the parts below which use the variable `name` should be renamed to `filename`. Thus if no --name is provided, the base_name of the filename is used as name, otherwise the explicitly defined name. This should also be consistent throughout the protocols, as now the name is not necessarily the filename anymore.

bari12 commented on lib/rucio/rse/protocols/protocol.py:
```diff
@@ -65,7 +65,8 @@ def lfns2pfns(self, lfns):
         lfns = [lfns] if type(lfns) == dict else lfns
         for lfn in lfns:
-            scope, name = lfn['scope'], lfn['name']
+            print "generating lfn for", lfn
```
Remove print

bari12 commented on lib/rucio/rse/protocols/protocol.py:
```diff
@@ -99,7 +100,8 @@ def __lfns2pfns_client(self, lfns):
             scope = lfn['scope']
-            name = lfn['name']
+            name = lfn.get('lfn', lfn['name'])
+            print "Processing LFN", lfn
```
Remove print

bari12 commented on lib/rucio/rse/protocols/protocol.py:
```diff
@@ -45,7 +45,7 @@ def __init__(self, protocol_attr, rse_settings):
                 setattr(self, '_get_path', self._get_path_nondeterministic_server)
         else:
-            self.attributes['determinism_type'] = 'default'
+            self.attributes['determinism_type'] = self.rse.get('determinism_type', 'default')
```
Need to check if `determinsm_type` is part of the rse_settings. I am not sure.

vingar replied:
it's deterministic, it is defined on rse and should be in rse_seetings

bari12 replied:
Yes, but deterministic is a boolean flag of the RSE. So for this you would need to have both the flag plus the determinism_type.

bari12 commented on lib/rucio/rse/protocols/protocol.py:
```diff
@@ -120,7 +122,10 @@ def _get_path(self, scope, name):
         if scope.startswith('user') or scope.startswith('group'):
             scope = scope.replace('.', '/')
-        return '%s/%s/%s/%s' % (scope, hstr[0:2], hstr[2:4], name)
+        if self.attributes.get('determinism_type', 'default') == 'identity':
```
For this part I really wonder if it would be beneficial if it comes from the policy and not the determinism_type from the RSE. e.g. for CMS the 'default' policy could be defined as `'%s/%s' % (scope, name)` while for ATLAS it is the hash part. Both have advantages, but I am mostly wondering if you forget to set the determinism_type for a newly created RSE you will get the hashed pathes until you fix it, while by setting the default via the policy that won't happen. Any preferences?

bbockelm commented:
Hm.  For some reason, GitHub didn't automatically link this to #377.  This comment should do the trick..

bari12 commented:
Now that everyone is back in the office, coming back to this.

I think what still should be done is:
- Remove the prints
- `lfn` should be renamed to `name`, so that the CLI parameter is `--name`. This is more consistent with Rucio terminology. Thus internally, everywhere where we use the variable `name` now, it has to be renamed to `filename`. If `--name` is not given, the `base_name` of the file is used as `name`.
- `self.rse.get('determinism_type', 'default')` will not work, as determinism_type is not part of `core.rse.get_rse_protocols` return value. This has to be added too.
  
Otherwise I think the patch is ready.

bari12 committed:
Merge pull request #434 from bbockelm/patch-433-client_avoid_session_reuse

client: Recreate session after authorization failure. Fix #433

bari12 committed:
Merge pull request #437 from bbockelm/patch-436-client_escape_urls

client: Escape DIDs and scope embedded in URLs.  Fix #436

bari12 committed:
Merge pull request #440 from mlassnig/patch-439-database__bootstrap_mismatch_between_SSH_and_GSS

database: bootstrap mismatch between SSH and GSS

bbockelm committed:
Minimal commit to expose `name` instead of `lfn`.

In Rucio terminology, `lfn` is supposed to be `scope:name` which is
inconsistent with how this patch series referred to it.

Switch the CLI to use the Rucio terminology and do the minimal
fixups in `bin/rucio`.

bbockelm committed:
Allow more explicit control of the LFN and PFN.

This allows the upload path to track the "file name" (on local disk)
and the LFN separately, allowing the `rucio upload` client to expose
the option of explicitly providing a LFN that differs from the file
name.

Additionally, to help upload files with '/' in them that are effectively
pre-hashed, this allows the RSE to be marked as having the 'identity'
determinism type, meaning to skipping the hashing step of generating
a PFN.

bbockelm committed:
Add new attribute name for LFN2PFN algorithm.

Previously, `determinism_type` was overloaded to specify the LFN2PFN
algorithm to use.  I decided this was a bad idea -- too hard to
identify what else might be using `determinism_type`.  Now, the
RSE key used is called `lfn2pfn_algorithm`.

Additionally, the value of this RSE attribute is now available
to the protocols.

The intent is that the available algorithms in core Rucio will be
simple / naive -- any complex mapping of LFN2PFN (such as the full
CMS case) will probably need to become a plugin mechanism.

bbockelm committed:
More consistent use of name and filename.

Now, we should (mostly) utilize `name` as part of the DID and
`filename` as the name of the file in the local directory.

bbockelm commented:
(note: I just rebased the branch in order to maintain sanity)

bbockelm commented:
Travis failures appear unrelated?

bari12 commented on bin/rucio:
```diff
@@ -946,17 +960,18 @@ def upload(args):
+                current_file_info['meta']['guid'] = generate_uuid()
+            list_files.append(current_file_info)
+            directory_worklist = lfns.setdefault(dir_name, [])
+            # TODO: why is this different from current_file_info?
```
I don't know. To me it looks like we do not need the `directory_worklist` at all?

bari12 commented on bin/rucio:
```diff
@@ -1003,13 +1018,28 @@ def upload(args):
     else:
         logger.debug('Skipping dataset registration')
 
+    def make_valid_did(lfn_dict):
```
Do we still need the `make_valid_did` method? Also in the rsemanager? It looks like to me that the file dictionary now always consists of `scope`, `name` and `filename`, thus it is not needed anymore?

bbockelm replied:
`filename` isn't a valid key in the JSON schema for a DID.  Hence, if we don't remove it, it fails the JSON validation later...

bari12 replied:
Right, understood. In that case I would suggest we add it to rucio.common.utils, so the code doesn't need to be duplicated. Then the patch is ready from my side 🎉  🕺 🎆

bbockelm committed:
Reuse any existing session.

If the caller provides a session, utilize that for the attribute
lookup.  Otherwise, a new session is always made - potentially causing
significant churn.

bbockelm committed:
Remove dead code.

bbockelm committed:
Add common implementation of make_valid_dids to utils.

bbockelm commented:
Alright, I think this one is ready to roll!

mlassnig reviewed (commented):
great! thanks a lot, brian! i think this is ready, but i leave the final signoff to @bari12

vingar reviewed (approved):
phew what a pull request ! all comments from Martin have been implemented and looks fine for me.

bbockelm commented:
Indeed, in retrospect, it conflated two issues that could have been made into separate PRs.  Hopefully can avoid that next time!
